### DISCUSSION TITLE
Using with Factory-boy?

### DISCUSSION THREAD

**Discussion by rsb177** [collerek/ormar#347] [Q&A]
Just curious if anybody has used Ormar w/ factory-boy?  I was trying to use factory-boy's `SQLAlchemyModelFactory`, and tweaking it so it would work, but Ii'm currently erroring out saying that my model is not mapped.  I've tried connecting it to the metadata, but I keep getting told that the table is already defined.  So I figure before I dive too deep into this, has anybody else tried to use factory-boy with Ormar?

**Comment by mciszczon** [comment]
I was tinkering with `factory_boy` as well and started to question myself... Do we even need it, when we can _almost_ do the same using just `pydantic`/`ormar`?

My initial idea is to make use of https://github.com/Goldziher/pydantic-factories to simplify my life a bit (I don' have to care about handling defaults or faking values). The usage would be simple, you just create a factory inheriting from `pydantic_factories.ModelFactory`, which will produce you a nice mock, which also has got the `.save()` method on it to save it to the database.

You could even create a `CustomModelFactory` class, something like this:

```python
class CustomModelFactory(pydantic_factories.ModelFactory):
	async def build(*args, **kwargs):
		mock = super().build(*args, **kwargs)
		await mock.save()
		return mock
```

then your mock would automatically get saved to the database. I will probably write something like this soon, as I'd need it too.

**Reply by mciszczon** [reply]
Actually, seems `pydantic_factories` already has it provided: https://github.com/Goldziher/pydantic-factories#persistence

**Reply by sondrelg** [reply]
I've been meaning to get around to implementing a proper ormar async-persistence extension to pydantic-factories for a while. Doesn't look like I'll get the chance anytime soon, but think that would be a huge asset. In the meantime I just do this:

```
obj = MyFactory.build(...)
obj.save()
```

Works really well for me 👍

**Reply by mciszczon** [reply]
@sondrelg 

Just done this and it works flawlessly as far as I can tell:

```python
from pydantic_factories import OrmarModelFactory


class Factory(OrmarModelFactory):
    @classmethod
    async def create(cls, **kwargs):
        built_object = cls.build(**kwargs)
        await built_object.save()
        return built_object

    @classmethod
    async def create_batch(cls, size: int, **kwargs):
        built_objects = cls.batch(size, **kwargs)
        await cls._get_model().objects.bulk_create(built_objects)
        return built_objects
```

Then, once you subclass this in your  factory, you can use create and create_batch instead of build and build_batch, they have the same interfaces:

```python
class SomeModelFactory(Factory):
	__model__ = SomeModel


some_model_instance: List[SomeModel] = await SomeModelFactory.create_batch(2)
```

I know it completely omits the Persistence Methods part from `pydantic_factories`, but the documentation there was unclear to me and this little piece of code does this in a much more concise way anyway IMHO.

**Reply by sondrelg** [reply]
Sweet, I'll try this out today 🎉  Looks good!

**Reply by sondrelg** [reply]
I've run into some issues with optionality, and the `OrmarModelFactory` seems to have some kinks to iron out. Hopefully we can get to the bottom of it and submit a PR, but this is what I've ended up with for now:

```python
T = TypeVar('T', bound=Model)


class Factory(ModelFactory, Generic[T]):
    __model__: type[T]
    __allow_none_optionals__ = False

    @classmethod
    async def create(cls, **kwargs: Any) -> T:
        built_object = cls.build(**kwargs)
        await built_object.save()
        return built_object

    @classmethod
    async def create_batch(cls, size: int, **kwargs) -> list[T]:
        built_objects = cls.batch(size, **kwargs)
        await cls._get_model().objects.bulk_create(built_objects)
        return built_objects
```

This seems to work for me, and the `TypeVar` addition solves mypy errors 👍

**Reply by mciszczon** [reply]
I also had a problem, check this issue: https://github.com/Goldziher/pydantic-factories/issues/27 I'll make a fix PR later today.

Thanks for ironing out the code and adding types!
